### PR TITLE
*: Fix issue #827

### PR DESCRIPTION
**Pull request by zimulala**
Fix: https://github.com/pingcap/tidb/issues/827

### CODE REVIEW

File: executor/executor_write.go [hunk 1]
```diff
@@ -637,7 +641,11 @@ func (e *InsertValues) fillRowData(cols []*column.Col, vals []types.Datum) ([]ty
 }
 
 func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struct{}) error {
-	var defaultValueCols []*column.Col
+	var (
+		pkHandleColOff   int
```
Comment (coocood) [inline executor/executor_write.go]: pkHandleCol already has Offset.

File: executor/executor_write.go [hunk 2]
```diff
@@ -637,7 +641,11 @@ func (e *InsertValues) fillRowData(cols []*column.Col, vals []types.Datum) ([]ty
...
+		pkHandleCol      *column.Col
```
Comment (coocood) [inline executor/executor_write.go]: I think `rewriteValueCol` is a better name.

File: executor/executor_write.go [hunk 3]
```diff
@@ -637,6 +641,7 @@ func (e *InsertValues) fillRowData(cols []*column.Col, vals []types.Datum) ([]ty
 }
 
 func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struct{}) error {
+	var pkHandleCol *column.Col
```
Comment (coocood) [inline executor/executor_write.go]: I think rename it to `rewriteValueCol` would be better. check `pkHandleCol`is set indicates there is pk handle column, but it is set only when it has allocated a record ID.

Comment (zimulala) [inline executor/executor_write.go]: gotcha

File: executor/executor_write.go [hunk 4]
```diff
@@ -662,6 +667,8 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struc
 				// `insert t (c1) values(1),(2),(3);`
 				// Last insert id will be 1, not 3.
 				variable.GetSessionVars(e.ctx).SetLastInsertID(uint64(recordID))
+				// It's used to retry.
+				rewriteValueCol = c
```
Comment (shenli) [inline executor/executor_write.go]: AutoIncrement column maybe not PrimaryKey. So should we move this outside the bracket?

File: executor/executor_write.go [hunk 5]
```diff
@@ -662,6 +667,8 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struc
 				// `insert t (c1) values(1),(2),(3);`
 				// Last insert id will be 1, not 3.
 				variable.GetSessionVars(e.ctx).SetLastInsertID(uint64(recordID))
+				// It's used to retry.
```
Comment (shenli) [inline executor/executor_write.go]: "used for retry" or "used to retry InsertStmt"

**Comment by shenli** [comment]
It is not easy to understand the code. Please add a few comments.

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**